src/runtime/program/precompile_programs/lib.zig

```diff
@@ -0,0 +1,175 @@
+const std = @import("std");
+const sig = @import("../../../sig.zig");
+
+pub const ed25519 = @import("ed25519.zig");
+pub const secp256k1 = @import("secp256k1.zig");
+pub const secp256r1 = @import("secp256r1.zig");
+
+const Pubkey = sig.core.Pubkey;
+const Ed25519 = std.crypto.sign.Ed25519;
+
+pub const ed25519Verify = ed25519.verify;
+pub const secp256k1Verify = secp256k1.verify;
+pub const secp256r1Verify = secp256r1.verify;
+
+pub const COMPUTE_UNITS = 0; // does this consume compute units?
+
+// TODO: should be moved to global features file
+pub const SECP256R1_FEATURE_ID =
+    Pubkey.parseBase58String("sr11RdZWgbHTHxSroPALe6zgaT5A1K9LcE4nfsZS4gi") catch unreachable;
+
+pub const PRECOMPILES = [_]Precompile{
+    .{
+        .program_id = sig.runtime.ids.PRECOMPILE_ED25519_PROGRAM_ID,
+        .function = ed25519Verify,
+        .required_feature = null,
+    },
+    .{
+        .program_id = sig.runtime.ids.PRECOMPILE_SECP256K1_PROGRAM_ID,
+        .function = secp256k1Verify,
+        .required_feature = null,
+    },
+    .{
+        .program_id = sig.runtime.ids.PRECOMPILE_SECP256R1_PROGRAM_ID,
+        .function = secp256r1Verify,
+        .required_feature = SECP256R1_FEATURE_ID,
+    },
+};
+
+pub fn verifyPrecompiles(
+    allocator: std.mem.Allocator,
+    transaction: sig.core.Transaction,
+    feature_set: sig.runtime.FeatureSet,
+) (PrecompileProgramError || error{OutOfMemory})!void {
+    // could remove this alloc by passing in the transaction in directly, but maybe less clean
+    var instruction_datas: ?[]const []const u8 = null;
+    defer if (instruction_datas) |instr_datas| allocator.free(instr_datas);
+
+    for (transaction.msg.instructions) |instruction| {
+        const program_id = transaction.msg.account_keys[instruction.program_index];
+        for (PRECOMPILES) |precompile| {
+            if (!precompile.program_id.equals(&program_id)) continue;
+
+            const precompile_feature_enabled = precompile.required_feature == null or
+                feature_set.active.contains(precompile.required_feature.?);
+            if (!precompile_feature_enabled) continue;
+
+            const datas = instruction_datas orelse blk: {
+                const buf = try allocator.alloc([]const u8, transaction.msg.instructions.len);
+                for (transaction.msg.instructions, 0..) |instr, i| buf[i] = instr.data;
+                instruction_datas = buf;
```

Comment:
dont we need to free the previous `instruction_datas` if its not null first, before setting it so we dont leak?

Reply:
This only allocates once - when instruction_datas is non-null, we just reuse it.